### PR TITLE
Update Cubeviz to work on data with spectral axis last

### PR DESCRIPTION
**Pull request by rosteen**
### Description

This implements compatibility with https://github.com/glue-viz/glue-astronomy/pull/68, which stops reordering Cube axes and leaves things the way they are in the `specutils.Spectrum1D` object. Opening as draft since this is very experimental and still needs work - I've fixed the data loading and slice plugin, but I'm sure a couple other things were broken by this. Mainly I wanted other people to see it if they're curious.

### TODO

- [x] Uncomment https://github.com/spacetelescope/jdaviz/blob/a9d2dfac2baae322719953bcf96e85fbc705bb78/tox.ini#L57
- [x] Change https://github.com/spacetelescope/jdaviz/blob/a9d2dfac2baae322719953bcf96e85fbc705bb78/setup.cfg#L36 to be `glue-astronomy>=0.4`.

xref #1190 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by rosteen** [comment]
The other motivation for posting this now is to make sure it doesn't conflict _too_ much with #1040 .

**Comment by pllim** [comment]
The history seems weird. Might wanna rebase against upstream main.

File: jdaviz/configs/cubeviz/plugins/parsers.py [hunk 1]
```diff
@@ -119,15 +119,24 @@ def _parse_hdu(app, hdulist, file_name=None):
         if hdu.data.dtype in (int, np.uint, np.uint32) or \
                 any(x in hdu.name.lower() for x in EXT_TYPES['mask']):
             app.add_data_to_viewer('mask-viewer', data_label)
+            viewer = app.get_viewer('mask-viewer')
+            viewer.state.x_att_world = app.data_collection[data_label].id['Right Ascension']
```
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:123]: Does this work with data with no WCS? I am trying to think how this would fit into #1040 .

Comment (rosteen) [reply jdaviz/configs/cubeviz/plugins/parsers.py:123]: I think I need a case here to fall back on `Pixel Axis 0 [z]` (or whichever is appropriate) for things with no WCS.

Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/parsers.py:123]: I am not sure which axis that would be. I have found that specutils only attempt to swapaxes if it detects a WCS, so for anything without WCS, I needed to transpose. If you look at PR 1040, you will see a lot of `array.T` happening here and there. 🤪  Do I still need that with this patch?

**Comment by rosteen** [comment]
CI won't pass yet because this needs https://github.com/glue-viz/glue-astronomy/pull/68. The tests pass locally on my machine with that installed.

**Comment by pllim** [comment]
Upstream PR is merged, so theoretically, if you pin `glue-astronomy` to dev version in `setup.cfg`, your CI would pass. I also added a TODO section in your original post above because your PR should undo two things added in #1190 . FYI.

**Comment by rosteen** [comment]
I think this is almost ready for in-depth review. There's one persistent bug that I'm aware of that I haven't fixed yet, where calculating a moment with a viewer selected to auto-plot the results will throw an `IncompatibleDataException` error from Glue `compute_fixed_resolution_buffer` and crash the app (note that it works fine if you calculate the moment, then go into the data selection menu for the viewer and change to the moment data there). I need to log off for the day, I wonder if you might have some insight on that @astrofrog? To reproduce, run the `CubevizExample` notebook and go to the `Moment Maps` plugin, select a viewer (I recommend Center or Right) and hit `Calculate`. Reminder that this PR requires dev `glue-astronomy`.

**Comment by rosteen** [comment]
I fixed the above error with a small change to `app.py` to clear existing data in a viewer first before selecting the new data, rather than doing both in one step.

**Comment by pllim** [comment]
We need to pin to glue-astronomy dev with this PR, right?

**Comment by rosteen** [comment]
> We need to pin to glue-astronomy dev with this PR, right?

Yes, the setup.cfg file here has it pinned.

**Comment by astrofrog** [comment]
@rosteen - did you fix the issue you asked me about above or should I still investigate?

**Comment by rosteen** [comment]
> @rosteen - did you fix the issue you asked me about above or should I still investigate?

I did figure out a fix, no need to investigate.

File: jdaviz/configs/cubeviz/helper.py [hunk 2]
```diff
@@ -16,6 +18,21 @@ def __init__(self, *args, **kwargs):
...
+        self.app.hub.subscribe(self, AddDataMessage,
+                               handler=self._set_spectrum_x_axis)
+
+    def _set_spectrum_x_axis(self, msg):
+        if msg.viewer_id != "cubeviz-3":
+            return
+        viewer = self.app.get_viewer("spectrum-viewer")
+        ref_data = viewer.state.reference_data
+        if ref_data and ref_data.ndim == 3:
+            for att_name in ["Wave", "Wavelength", "Freq", "Frequency"]:
```
Comment (pllim) [inline jdaviz/configs/cubeviz/helper.py:30]: Do people make cubes in wave number, etc? How do these map to supported CTYPE and how many of those we need to actually support?

https://github.com/astropy/astropy/blob/43ee5806e9c6f7d58c12c1cb9287b3c61abe489d/astropy/wcs/wcsapi/fitswcs.py#L88

File: jdaviz/configs/default/plugins/line_lists/line_lists.py [hunk 3]
```diff
@@ -379,7 +379,10 @@ def _auto_slider_range(self, event=None):
         ndec = -np.log10(half_range)
         if ndec > 0:
             # round to at least 2 digits, or the first significant digit
-            ndec = np.max([2, int(np.ceil(ndec))])
+            try:
+                ndec = np.max([2, int(np.ceil(ndec))])
```
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:383]: I feel like @kecnry already fixed this in #1221 but in a different way.

File: jdaviz/configs/default/plugins/line_lists/line_lists.py [hunk 4]
```diff
@@ -413,7 +416,10 @@ def _on_rs_slider_step_updated(self, event):
             ndec = np.max([2, np.ceil(ndec)+1])
         else:
             ndec = 1
-        self.rs_slider_ndigits = int(ndec)
+        try:
+            self.rs_slider_ndigits = int(ndec)
+        except OverflowError:
```
Comment (pllim) [inline jdaviz/configs/default/plugins/line_lists/line_lists.py:421]: See https://github.com/spacetelescope/jdaviz/pull/1174/files#r844136625

File: jdaviz/tests/test_subsets.py [hunk 5]
```diff
@@ -60,12 +60,14 @@ def test_region_from_subset_3d(jdaviz_app):
...
 def test_region_from_subset_profile(jdaviz_app, spectral_cube_wcs):
-    data = Data(flux=np.ones((256, 128, 128)), label='Test 1D Flux', coords=spectral_cube_wcs)
+    data = Data(flux=np.ones((128, 128, 256)), label='Test 1D Flux', coords=spectral_cube_wcs)
     jdaviz_app.data_collection.append(data)
 
     jdaviz_app.add_data_to_viewer('spectrum-viewer', 'Test 1D Flux')
 
-    jdaviz_app.get_viewer("spectrum-viewer").apply_roi(XRangeROI(5, 15.5))
+    viewer = jdaviz_app.get_viewer('spectrum-viewer')
+    viewer.state.x_att = jdaviz_app.data_collection['Test 1D Flux'].id['Frequency']
```
Comment (pllim) [inline jdaviz/tests/test_subsets.py:69]: What does it mean that you need to add this line to the test from a dev perspective?

File: jdaviz/configs/default/plugins/line_lists/line_lists.py (hunk 3 above)
Comment (rosteen) [reply jdaviz/configs/default/plugins/line_lists/line_lists.py:383]: Good call, I'll revert this fix in favor of that one.

File: jdaviz/tests/test_subsets.py (hunk 5 above)
Comment (rosteen) [reply jdaviz/tests/test_subsets.py:69]: I actually might not need this now that I moved the relevant logic from the parser to the helper, let me check...

Comment (rosteen) [reply jdaviz/tests/test_subsets.py:69]: Ah, right. The answer is that changing the x-axis of the `spectrum-viewer` automatically gets triggered via `AddDataMessage` in the Cubeviz helper class, but these tests aren't using the Cubeviz helper class, they're initializing things at a lower level via jdaviz.Application. I think the best thing to do here is to update to test using the helper class, to more accurately represent actual use cases.

File: jdaviz/configs/cubeviz/helper.py (hunk 2 above)
Comment (rosteen) [reply jdaviz/configs/cubeviz/helper.py:30]: Looking at the glue-astronomy translator, it seems like the other likely options are Wavenumber, Velocity and Energy. I'll add those.

File: jdaviz/configs/cubeviz/plugins/viewers.py [hunk 6]
```diff
@@ -35,6 +35,20 @@ class CubevizImageView(BqplotImageView, JdavizViewerMixin):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self._initialize_toolbar_nested()
+        self.state.add_callback('reference_data', self._initial_x_axis)
+
+    def _initial_x_axis(self, *args):
+        # Make sure that the x_att is correct on data load
+        ref_data = self.state.reference_data
+        if ref_data and ref_data.ndim == 3:
+            for att_name in ["Right Ascension", "RA"]:
```
[the remark this answers is not in the document]
Comment (rosteen) [reply jdaviz/configs/cubeviz/plugins/viewers.py:44]: Digging into Glue and looking at what the other options that might occur here are, it seems like Galactic Longitude/Latitude are probably the only others we might need to worry about (I don't want to mess with supporting solar people and their crazy helioprojective things right now 😆). I don't know if people would want lat or lon on the x-axis, I'll ask the POs.

**Comment by javerbukh** [comment]
Got an error while testing the plugins in cubeviz. I pretty much just opened each plugin in order and ran it, and when I opened the line analysis plugin, I saw the following traceback

<details>

```

---------------------------------------------------------------------------
IncompatibleAttribute                     Traceback (most recent call last)
File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/data.py:1385, in Data.get_mask(self, subset_state, view)
   1384 try:
-> 1385     return subset_state.to_mask(self, view=view)
   1386 except IncompatibleAttribute:

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/subset.py:773, in RangeSubsetState.to_mask(self, data, view)
    771 @contract(data='isinstance(Data)', view='array_view')
    772 def to_mask(self, data, view=None):
--> 773     x = data[self.att, view]
    774     result = (x >= self.lo) & (x <= self.hi)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/data.py:571, in BaseCartesianData.__getitem__(self, key)
    569         raise IncompatibleAttribute(_k)
--> 571 return self.get_data(key, view=view)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/data.py:1361, in Data.get_data(self, cid, view)
   1360 else:
-> 1361     raise IncompatibleAttribute(cid)
   1363 if view is not None:

IncompatibleAttribute: Wave

During handling of the above exception, another exception occurred:

IncompatibleAttribute                     Traceback (most recent call last)
File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:756, in Widget._handle_msg(self, msg)
    754         if 'buffer_paths' in data:
    755             _put_buffers(state, data['buffer_paths'], msg['buffers'])
--> 756         self.set_state(state)
    758 # Handle a state request.
    759 elif method == 'request_state':

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:625, in Widget.set_state(self, sync_data)
    622 if name in self.keys:
    623     from_json = self.trait_metadata(name, 'from_json',
    624                                     self._trait_from_json)
--> 625     self.set_trait(name, from_json(sync_data[name], self))

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue_jupyter/state_traitlets_helpers.py:151, in GlueState.update_state_from_json(self, json, widget)
    149 self._block_on_state_change = True
    150 try:
--> 151     update_state_from_dict(state, json)
    152 finally:
    153     self._block_on_state_change = False

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue_jupyter/state_traitlets_helpers.py:70, in update_state_from_dict(state, changes)
     68     setattr(state, name, get_cmap(changes[name]))
     69 else:
---> 70     setattr(state, name, changes[name])

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/echo/core.py:261, in HasCallbackProperties.__setattr__(self, attribute, value)
    260 def __setattr__(self, attribute, value):
--> 261     super(HasCallbackProperties, self).__setattr__(attribute, value)
    262     if self.is_callback_property(attribute):
    263         self._notify_global(**{attribute: value})

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/echo/core.py:76, in CallbackProperty.__set__(self, instance, value)
     74 new = self.__get__(instance)
     75 if old != new:
---> 76     self.notify(instance, old, new)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/echo/core.py:125, in CallbackProperty.notify(self, instance, old, new)
    123     return
    124 for cback in self._callbacks.get(instance, []):
--> 125     cback(new)
    126 for cback in self._2arg_callbacks.get(instance, []):
    127     cback(old, new)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/core/template_mixin.py:86, in PluginTemplateMixin._mxn_update_plugin_opened(self, new_value)
     84 app_state = self.app.state
     85 tray_names_open = [app_state.tray_items[i]['name'] for i in app_state.tray_items_open]
---> 86 self.plugin_opened = app_state.drawer and self._registry_name in tray_names_open

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/traitlets/traitlets.py:606, in TraitType.__set__(self, obj, value)
    604     raise TraitError('The "%s" trait is read-only.' % self.name)
    605 else:
--> 606     self.set(obj, value)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/traitlets/traitlets.py:595, in TraitType.set(self, obj, value)
    591     silent = False
    592 if silent is not True:
    593     # we explicitly compare silent to True just in case the equality
    594     # comparison above returns something other than True/False
--> 595     obj._notify_trait(self.name, old_value, new_value)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/traitlets/traitlets.py:1219, in HasTraits._notify_trait(self, name, old_value, new_value)
   1218 def _notify_trait(self, name, old_value, new_value):
-> 1219     self.notify_change(Bunch(
   1220         name=name,
   1221         old=old_value,
   1222         new=new_value,
   1223         owner=self,
   1224         type='change',
   1225     ))

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/ipywidgets/widgets/widget.py:686, in Widget.notify_change(self, change)
    683     if name in self.keys and self._should_send_property(name, getattr(self, name)):
    684         # Send new state to front-end
    685         self.send_state(key=name)
--> 686 super(Widget, self).notify_change(change)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/traitlets/traitlets.py:1229, in HasTraits.notify_change(self, change)
   1227 def notify_change(self, change):
   1228     """Notify observers of a change event"""
-> 1229     return self._notify_observers(change)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/traitlets/traitlets.py:1266, in HasTraits._notify_observers(self, event)
   1263 elif isinstance(c, EventHandler) and c.name is not None:
   1264     c = getattr(self, c.name)
-> 1266 c(event)

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:101, in LineAnalysis._on_plugin_opened_changed(self, *args)
     99     mark.visible = self.plugin_opened
    100 if self.plugin_opened:
--> 101     self._calculate_statistics()

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/configs/specviz/plugins/line_analysis/line_analysis.py:185, in LineAnalysis._calculate_statistics(self, *args, **kwargs)
    182     return
    184 if self.spectral_subset_selected != "Surrounding":
--> 185     sr = self.app.get_subsets_from_viewer("spectrum-viewer",
    186                                           subset_type="spectral").get(self.spectral_subset_selected) # noqa
    187 else:
    188     sr = None

File ~/Documents/jdaviz_dev/jdaviz/jdaviz/app.py:680, in Application.get_subsets_from_viewer(self, viewer_reference, data_label, subset_type)
    670 # There is a special case for 1d data (which is also not
    671 #  supported currently). We now eschew the use of the
    672 #  translation machinery entirely and construct the astropy
    673 #  region ourselves.
    674 elif value.data.ndim == 1:
    675     # Grab the data units from the glue-astronomy spectral axis
    676     # TODO: this needs to be much simpler; i.e. data units in
    677     #  the glue component objects
    678     # Cases where there is a single subset
    679     subregions_in_subset = _get_all_subregions(
--> 680             np.where(value.to_mask() == True)[0], # noqa
    681             value.data.coords.spectral_axis)
    683     regions[key] = subregions_in_subset
    684     continue

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/subset.py:182, in Subset.to_mask(self, view)
    169 @contract(view='array_view', returns='array')
    170 def to_mask(self, view=None):
    171     """
    172     Convert the current subset to a mask.
    173 
   (...)
    180 
    181     """
--> 182     return self.data.get_mask(self.subset_state, view=view)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/data.py:1387, in Data.get_mask(self, subset_state, view)
   1385     return subset_state.to_mask(self, view=view)
   1386 except IncompatibleAttribute:
-> 1387     return get_mask_with_key_joins(self, self._key_joins, subset_state, view=view)

File ~/opt/anaconda3/envs/jdaviz-rosteen-PR/lib/python3.8/site-packages/glue/core/joins.py:105, in get_mask_with_key_joins(data, key_joins, subset_state, view)
     99     else:
    101         raise Exception("Either the number of components in the key join sets "
    102                         "should match, or one of the component sets should ",
    103                         "contain a single component.")
--> 105 raise IncompatibleAttribute

IncompatibleAttribute: 
```

</details>

**Comment by javerbukh** [comment]
I can get more specific about how to reproduce:
1) Run cubeviz example notebook
2) Create a spatial subset and a spectral subset
3) Create a model that is a first order polynomial
4) Fit model, show in middle viewer, and apply to cube
5) Open line analysis and get traceback

Cubeviz still runs but the line analysis plugin hangs.

**Comment by javerbukh** [comment]
This is what I see on main. I do not see the same error as I saw on this branch.
Screen Shot 2022-04-12 at 12 23 29 PM

File: jdaviz/configs/cubeviz/plugins/slice/slice.py [hunk 7]
```diff
@@ -55,14 +55,19 @@ def _on_linked_changed(self, event):
                                           self._viewer_slices_changed)
 
     def _on_data_added(self, msg):
-        if len(msg.data.shape) == 3 and isinstance(msg.viewer, BqplotImageView):
-            self.max_value = msg.data.shape[0] - 1
+        if isinstance(msg.viewer, BqplotImageView):
+            if len(msg.data.shape) == 3:
+                self.max_value = msg.data.shape[-1] - 1
```
Comment (kecnry) [inline jdaviz/configs/cubeviz/plugins/slice/slice.py:60]: for the example notebook, this seems to incorrectly set a maximum of 73 slices (the slider does not allow scrubbing through the whole cube).

File: jdaviz/configs/default/plugins/gaussian_smooth/gaussian_smooth.py [hunk 8]
```diff
@@ -54,7 +54,8 @@ def _on_data_selected(self, event):
             return
 
         # NOTE: if this is ever used anywhere else, it should be moved into DatasetSelect
-        self.selected_data_is_1d = len(self.dataset.selected_dc_item.data.shape) == 1
+        if self.dataset.selected_dc_item is not None:
+            self.selected_data_is_1d = len(self.dataset.selected_dc_item.data.shape) == 1
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/gaussian_smooth/gaussian_smooth.py:58]: what scenario was causing this to trigger when it was still None?  (I'm just wondering if we need to handle defaulting `self.selected_data_is_1d` in this case, whereas the logic now will just leave it at its current value).

File: jdaviz/configs/cubeviz/plugins/slice/slice.py (hunk 7 above)
Comment (rosteen) [reply jdaviz/configs/cubeviz/plugins/slice/slice.py:60]: Do you have the required `glue-astronomy` version? The slice plugin works fine for me.

Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/slice/slice.py:60]: doh!  Forgot I had downgraded to test something else recently.  Sorry about that!

File: jdaviz/configs/default/plugins/gaussian_smooth/gaussian_smooth.py (hunk 8 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/gaussian_smooth/gaussian_smooth.py:58]: ok... that's probably a case that should be handled better across all plugins, but I think this is fine as-is for now

**Comment by rosteen** [comment]
See https://github.com/spacetelescope/jdaviz/issues/1245 for the followup on the Cubeviz fitting bug.

**Review by kecnry — APPROVED** [review]
Everything seems to work now, and it'll definitely be good to get this change in and behind us!

**Review by javerbukh — APPROVED** [review]
Sounds good, thanks for opening the ticket!

**Pull request merged**